### ISSUE TITLE
Buffer overflow with "--footer-html" and "data" URLs

### ISSUE THREAD

**Issue by metagriffin** [wkhtmltopdf/wkhtmltopdf#2280]
When invoking wkhtmltopdf with a header/footer with a data-scheme URL, such as:

```
wkhtmltopdf --footer-html 'data:text/html;base64,PGh0bWw+VGVzdDwvaHRtbD4K' IN OUT
```

the footer contains randomly appended characters -- see the screenshot:

![screenshot](https://cloud.githubusercontent.com/assets/2552231/6985427/3aa8ddd2-da01-11e4-9581-8dc32219dd86.png)

(the base64 decoded text is: ```<html>Test</html>```)

**Comment by ashkulz** [comment]
Hmm, `--footer-html` is supposed to be used with actual files and not a `data:` URI ... still a bug, though.

**Comment by ashkulz** [comment]
Thanks for the catch! Now it will give the error message:
```
Error: --footer-html should be a URL and not a string containing HTML code.
Exit with code 1, due to unknown error.
```

**Comment by metagriffin** [comment]
this is a very unfortunate solution. use of the "data" scheme is perfectly valid, IMHO, and very powerful. i was hoping to use the --footer-html to be able to control the footer very precisely, but now i need a persistent location for the HTML, ie. a server, which in my particularly context i do not have access to since all content is generated on the fly (and there is no server).

is there any way that you can re-implement this to allow for that?

**Comment by ashkulz** [comment]
How would query string parameters be handled in that case if data URIs where allowed?

**Comment by metagriffin** [comment]
just like in a browser, they should be ignored, i.e. no special handling needs to be done for query strings. for example, you can try the following in your browser:

```
data:text/html;base64,PGh0bWw+VGVzdDwvaHRtbD4K?foo=bar
```

**Comment by ashkulz** [comment]
The moment you add a query string, it does not work in Firefox & Chrome ... so I guess they are not allowed for data URIs.

**Comment by metagriffin** [comment]
i just confirmed that, sorry! -- i had tried it with FF, first without the query-string, which worked, and then simply added the QS, which also appeared to work because it was just leaving the page unchanged. unfortunately, it does not work if you go directly to it with the QS.

interestingly, if you add a ``#`` before the ``?``, it works in FF (but not chrome).

perhaps it is the additional query strings that are being added automatically that are causing the garbage text?

**Comment by ashkulz** [comment]
Not sure I want to investigate it further ... as long as it gets caught beforehand :smile:

**Comment by ashkulz** [comment]
A preview for the next release [`0.12.3-dev-79ff51e`](http://wkhtmltopdf.org/downloads.html#testing) is available, which should contain the fix for this issue. Please report back if it is not solved with the above version.

*Please note that the above downloads will be removed after the `0.12.3` release*.

**Comment by ashkulz** [comment]
[`0.12.3`](http://wkhtmltopdf.org/downloads.html#stable) has been released, which should contain the fix for this issue. Please report back if it is not solved with the above version.

*Note that the preview downloads mentioned above have been removed*.

**Comment by tyuwan** [comment]
Has this issue already been resolved? It still doesn't allows me to use the data-scheme. 

0.12.3 still checking for header and footer containing html code.

            if (looksLikeHtmlAndNotAUrl(s.header.htmlUrl)) {
                emit out.error("--header-html should be a URL and not a string containing HTML code.");
                fail();
                return;
            }

**Comment by ashkulz** [comment]
@tyuwan: the fix was to disallow data URIs because of the overflow and the inability to use query string parameters (see the above discussion).

**Comment by tyuwan** [comment]
@ashkulz: Is there a way to is there any way that you can re-implement to allow data string OR URL? Since I normally generated all the content on the fly.

**Comment by ashkulz** [comment]
@tyuwan: not really, there would also be limitations due to maximum command-line length (as `wkhtmltopdf` is cross-platform). It is simply better to use temporary files which you delete after conversion.

**Comment by metagriffin** [comment]
@ashkulz: speaking for myself (and i suspect for @tyuwan as well), i politely disagree with you. yes, you are correct that in the situation where there is a lot of data, then using the command line is not viable. for all other situations (definitely mine and probably @tyuwan's), using the command line is not a problem. furthermore, in my case, i am generating the string in server memory, so needing to create a temporary file is a significant annoyance (i need to write a lot of extra code, ensure correct access, ensure that the file is correctly deleted, more unit tests, etc).

please do us a favor and remove the *artificial* restriction that you have put in place and fix the buffer overflow keeping us from using your tool the way that best works for us. thanks!

**Comment by ashkulz** [comment]
data URIs [do not support query strings](https://developer.mozilla.org/en-US/docs/Web/HTTP/data_URIs) -- so they cannot be used for the header/footer HTML, _because that usage requires query string parameters_. The buffer overflow is probably because of incomplete validation in the Qt/WebKit code, which is why additional validation was added in wkhtmltopdf.

So this not an artificial restriction, but a limitation of the spec -- which was also referred to above.

**Comment by metagriffin** [comment]
ok, since by definition the content provided in a "data" uri scheme cannot depend on the query strings (since it is static), it would make sense to not include the query strings if the URL uses the "data" scheme. this is (for me) the perfect solution: i can generate static headers/footers and don't need to create temporary files.

so, that would comply with the spec and may address the qt/webkit problem, if that is where the problem was. would that be possible?

**Comment by ashkulz** [comment]
That would introduce a strange limitation -- for data URIs only static headers/footers are supported. I'm pretty sure it would come as a surprise to most people who think of using this functionality, so I'm not in favor of implementing it...

**Comment by pawepaw** [comment]
Isn't it possible to create temp file as you do for body html? it's a strange limitation that you have to create temp file for footer/header but not for body.

**Comment by musedivision** [comment]
Hey @ashkulz  I am in agreement with those above, Could we implement something as @pawepaw has suggested? 

I am using wkhtmltopdf on a serverless function. While the url/ server responses for footers and headers are fine for a standard server implementation. For a serverless implementation I am having to do much extra work to get html for headers and footers into wkhtmltopdf ( which i have already statically generated )

For me the limitation you mention above, would work fine within my solution:
"That would introduce a strange limitation -- for data URIs only static headers/footers are supported. "



Thanks for your work, pat

**Comment by RA1449** [comment]
I followed the link to madnight's fix foot rendering seen above.  He has a live demo up too.  I used Postman to send JSON with encoded html as content and footer and included options as per his example.  For simplicity I used the same encoded html for content and footer, which, when decoded, spells out 'Test'.  

This was a JSON I sent
{
    "contents": "PGh0bWw+VGVzdDwvaHRtbD4K",
    "options": {
        "margin-top": "20",
        "margin-left": "20",
        "margin-right": "20",
        "margin-bottom": "30"
    },
    "footer": "PGh0bWw+VGVzdDwvaHRtbD4K"
}`

He includes tempfile module in his code and decodes base64 footer.
